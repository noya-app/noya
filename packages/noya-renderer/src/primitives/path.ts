import Sketch from '@sketch-hq/sketch-file-format-ts';
import { CanvasKit, Path, PathOp } from 'canvaskit';
import { distance, Rect } from 'noya-geometry';
import { Point } from 'noya-state';
import { windowsOf } from 'noya-utils';
import { parsePoint, stringifyPoint } from '../primitives';

/**
 * The radius of an edge should be less than half the length of that edge
 */
function getEdgeRadius(a: Point, b: Point, fixedRadius: number): number {
  const dist = distance(a, b);

  return dist <= fixedRadius * 2 ? dist / 2 : fixedRadius;
}

/**
 * The radius of a corner should be symmetric on both sides
 */
function getCornerRadius(
  before: Point,
  corner: Point,
  after: Point,
  fixedRadius: number,
): number {
  const beforeRadius = getEdgeRadius(corner, before, fixedRadius);
  const afterRadius = getEdgeRadius(corner, after, fixedRadius);
  return Math.min(beforeRadius, afterRadius);
}

// Implementation loosely follows:
// https://github.com/google/skia/blob/13d6c4f55db0d1429f889704ed9b2e48968ac7a3/src/effects/SkCornerPathEffect.cpp#L24
function getStep(
  a: Point,
  b: Point,
  radius: number,
): { step: Point; drawSegment: boolean } {
  let dist = distance(a, b);

  const step: Point = { x: b.x - a.x, y: b.y - a.y };

  // Since we clamp radius beforehand, the dist will never be < radius * 2, only equal
  if (dist <= radius * 2) {
    step.x /= 2;
    step.y /= 2;

    return { step, drawSegment: false };
  }

  step.x *= radius / dist;
  step.y *= radius / dist;

  return { step, drawSegment: true };
}

export type ParsedCurvePoint = Omit<
  Sketch.CurvePoint,
  'curveFrom' | 'curveTo' | 'point'
> & {
  curveFrom: Point;
  curveTo: Point;
  point: Point;
};

export function scalePoint(point: Point, { x, y, width, height }: Rect): Point {
  return { x: x + point.x * width, y: y + point.y * height };
}

export function unscalePoint(
  point: Point,
  { x, y, width, height }: Rect,
): Point {
  return { x: (point.x - x) / width, y: (point.y - y) / height };
}

export function parseCurvePoint(
  curvePoint: Sketch.CurvePoint,
  frame: Rect,
): ParsedCurvePoint {
  return {
    ...curvePoint,
    curveFrom: scalePoint(parsePoint(curvePoint.curveTo), frame),
    curveTo: scalePoint(parsePoint(curvePoint.curveFrom), frame),
    point: scalePoint(parsePoint(curvePoint.point), frame),
  };
}

export function unparseCurvePoint(
  curvePoint: ParsedCurvePoint,
  frame: Rect,
): Sketch.CurvePoint {
  return {
    ...curvePoint,
    curveFrom: stringifyPoint(unscalePoint(curvePoint.curveTo, frame)),
    curveTo: stringifyPoint(unscalePoint(curvePoint.curveFrom, frame)),
    point: stringifyPoint(unscalePoint(curvePoint.point, frame)),
  };
}

export function path(
  CanvasKit: CanvasKit,
  points: Sketch.CurvePoint[],
  frame: Rect,
): Path {
  const curvePoints = points.map((point) => parseCurvePoint(point, frame));

  const pairs = windowsOf(curvePoints, 3, true);

  const path = new CanvasKit.Path();

  // Move to the starting point
  if (pairs[pairs.length - 1]) {
    const [prev, current, next] = pairs[pairs.length - 1];
    const currentPoint = current.point;
    const nextPoint = next.point;
    const prevPoint = prev.point;

    if (next.curveMode === Sketch.CurveMode.Straight) {
      const radius = getCornerRadius(
        prevPoint,
        currentPoint,
        nextPoint,
        current.cornerRadius,
      );

      const { step } = getStep(currentPoint, nextPoint, radius);

      path.moveTo(currentPoint.x + step.x, currentPoint.y + step.y);
    } else {
      path.moveTo(nextPoint.x, nextPoint.y);
    }
  }

  pairs.forEach((pair) => {
    const [prev, current, next] = pair;

    const currentPoint = current.point;
    const currentCurveTo = current.curveTo;
    const nextCurveFrom = next.curveFrom;
    const nextPoint = next.point;
    const prevPoint = prev.point;

    if (next.curveMode === Sketch.CurveMode.Straight) {
      const radius = getCornerRadius(
        prevPoint,
        currentPoint,
        nextPoint,
        current.cornerRadius,
      );

      const { step, drawSegment } = getStep(currentPoint, nextPoint, radius);

      if (current.cornerRadius > 0) {
        path.arcToTangent(
          currentPoint.x,
          currentPoint.y,
          currentPoint.x + step.x,
          currentPoint.y + step.y,
          radius,
        );
      } else {
        path.lineTo(currentPoint.x + step.x, currentPoint.y + step.y);
      }

      if (drawSegment) {
        path.lineTo(nextPoint.x - step.x, nextPoint.y - step.y);
      }
    } else {
      path.cubicTo(
        currentCurveTo.x,
        currentCurveTo.y,
        nextCurveFrom.x,
        nextCurveFrom.y,
        nextPoint.x,
        nextPoint.y,
      );
    }
  });

  path.close();

  // console.log(path.toCmds());

  return path;
}

const getBorderPositionPathOp = (
  CanvasKit: CanvasKit,
  position: Sketch.BorderPosition,
): PathOp | undefined => {
  switch (position) {
    case Sketch.BorderPosition.Center:
      return;
    case Sketch.BorderPosition.Outside:
      return CanvasKit.PathOp.Difference;
    case Sketch.BorderPosition.Inside:
      return CanvasKit.PathOp.Intersect;
  }
};

export function getStrokedPath(
  CanvasKit: CanvasKit,
  path: Path,
  width: number,
  pathOp?: PathOp,
) {
  const copy = path.copy();

  // A 0-width path will never change the original when unioned
  if (width === 0 && pathOp === CanvasKit.PathOp.Union) return copy;

  if (
    !copy.stroke({
      width,
    })
  ) {
    console.info('[getStrokedPath] Failed to stroke path');
    return copy;
  }

  if (pathOp !== undefined) {
    if (!copy.op(path, pathOp)) {
      console.info('[getStrokedPath] Failed to combine paths');
    }
  }

  return copy;
}

export function getStrokedBorderPath(
  CanvasKit: CanvasKit,
  path: Path,
  borderWidth: number,
  borderPosition: Sketch.BorderPosition,
) {
  switch (borderPosition) {
    case Sketch.BorderPosition.Center:
      break;
    case Sketch.BorderPosition.Outside:
    case Sketch.BorderPosition.Inside:
      borderWidth *= 2;
  }

  const pathOp = getBorderPositionPathOp(CanvasKit, borderPosition);

  return getStrokedPath(CanvasKit, path, borderWidth, pathOp);
}
