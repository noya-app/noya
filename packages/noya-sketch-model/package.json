{
  "name": "noya-sketch-model",
  "version": "1.0.0",
  "main": "src/index.ts",
  "license": "UNLICENSED"
}
