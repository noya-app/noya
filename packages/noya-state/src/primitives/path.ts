import { Sketch } from '@noya-app/noya-file-format';
import { Point, Rect, distance } from '@noya-app/noya-geometry';
import { PointString, SketchModel } from '@noya-app/noya-sketch-model';
import { clamp, rotate, windowsOf, zip } from '@noya-app/noya-utils';
import { CanvasKit, Path, PathOp } from 'canvaskit';
import {
  CommandWithoutQuadratics,
  makePathsFromCommands,
} from 'noya-import-svg';
import {
  lineCapStyle,
  lineJoinStyle,
  parsePoint,
  stringifyPoint,
} from 'noya-state';
import { PathCommand, PathCommandVerb, parsePathCmds } from './pathCommand';

/**
 * The radius of an edge should be less than half the length of that edge
 */
function getEdgeRadius(a: Point, b: Point, cornerRadius: number): number {
  const dist = distance(a, b);

  return dist <= cornerRadius * 2 ? dist / 2 : cornerRadius;
}

/**
 * The radius of a corner should be symmetric on both sides
 */
function getCornerRadius(
  before: Point,
  corner: Point,
  after: Point,
  cornerRadius: number,
): number {
  const beforeRadius = getEdgeRadius(corner, before, cornerRadius);
  const afterRadius = getEdgeRadius(corner, after, cornerRadius);
  return Math.min(beforeRadius, afterRadius);
}

// Implementation loosely follows:
// https://github.com/google/skia/blob/13d6c4f55db0d1429f889704ed9b2e48968ac7a3/src/effects/SkCornerPathEffect.cpp#L24
function getStep(
  a: Point,
  b: Point,
  radius: number,
): { step: Point; drawSegment: boolean } {
  let dist = distance(a, b);

  const step: Point = { x: b.x - a.x, y: b.y - a.y };

  // Since we clamp radius beforehand, the dist will never be < radius * 2, only equal
  if (dist <= radius * 2) {
    step.x /= 2;
    step.y /= 2;

    return { step, drawSegment: false };
  }

  step.x *= radius / dist;
  step.y *= radius / dist;

  return { step, drawSegment: true };
}

export type DecodedCurvePoint = Omit<
  Sketch.CurvePoint,
  'curveFrom' | 'curveTo' | 'point'
> & {
  curveFrom: Point;
  curveTo: Point;
  point: Point;
};

export function scalePoint(point: Point, { x, y, width, height }: Rect): Point {
  return { x: x + point.x * width, y: y + point.y * height };
}

export function unscalePoint(
  point: Point,
  { x, y, width, height }: Rect,
): Point {
  return { x: (point.x - x) / width, y: (point.y - y) / height };
}

export function decodeCurvePoint(
  curvePoint: Sketch.CurvePoint,
  frame: Rect,
): DecodedCurvePoint {
  return {
    ...curvePoint,
    hasCurveFrom: curvePoint.hasCurveTo,
    hasCurveTo: curvePoint.hasCurveFrom,
    curveFrom: scalePoint(parsePoint(curvePoint.curveTo), frame),
    curveTo: scalePoint(parsePoint(curvePoint.curveFrom), frame),
    point: scalePoint(parsePoint(curvePoint.point), frame),
  };
}

export function encodeCurvePoint(
  curvePoint: DecodedCurvePoint,
  frame: Rect,
): Sketch.CurvePoint {
  return {
    ...curvePoint,
    hasCurveFrom: curvePoint.hasCurveTo,
    hasCurveTo: curvePoint.hasCurveFrom,
    curveFrom: stringifyPoint(unscalePoint(curvePoint.curveTo, frame)),
    curveTo: stringifyPoint(unscalePoint(curvePoint.curveFrom, frame)),
    point: stringifyPoint(unscalePoint(curvePoint.point, frame)),
  };
}

function curveTo(
  path: Path,
  current: DecodedCurvePoint,
  next: DecodedCurvePoint,
) {
  const curveTo = current.hasCurveTo ? current.curveTo : current.point;
  const curveFrom = next.hasCurveFrom ? next.curveFrom : next.point;
  const nextPoint = next.point;

  path.cubicTo(
    curveTo.x,
    curveTo.y,
    curveFrom.x,
    curveFrom.y,
    nextPoint.x,
    nextPoint.y,
  );
}

function lineTo(path: Path, next: DecodedCurvePoint) {
  const nextPoint = next.point;

  path.lineTo(nextPoint.x, nextPoint.y);
}

function curveOrLineTo(
  path: Path,
  current: DecodedCurvePoint,
  next: DecodedCurvePoint,
) {
  if (
    current.curveMode === Sketch.CurveMode.Straight &&
    next.curveMode === Sketch.CurveMode.Straight
  ) {
    lineTo(path, next);
  } else {
    curveTo(path, current, next);
  }
}

type CurveInfo = {
  radius: number;
  step: Point;
  drawSegment: boolean;
};

function stepToCurve(
  path: Path,
  current: DecodedCurvePoint,
  curveInfo: CurveInfo,
) {
  const currentPoint = current.point;

  const { radius, step } = curveInfo;

  if (radius > 0) {
    path.arcToTangent(
      currentPoint.x,
      currentPoint.y,
      currentPoint.x + step.x,
      currentPoint.y + step.y,
      radius,
    );
  } else {
    path.lineTo(currentPoint.x + step.x, currentPoint.y + step.y);
  }
}

export function path(
  CanvasKit: CanvasKit,
  points: Sketch.CurvePoint[],
  frame: Rect,
  isClosed: boolean,
): Path {
  const path = new CanvasKit.Path();

  if (points.length < 2) return path;

  const curvePoints = points.map((point) => decodeCurvePoint(point, frame));

  // Curves with only 2 points don't support corner radius, so we handle this
  // case specially
  if (curvePoints.length === 2) {
    const [current, next] = curvePoints;

    path.moveTo(current.point.x, current.point.y);

    curveOrLineTo(path, current, next);

    if (isClosed) {
      curveOrLineTo(path, next, current);

      path.close();
    }

    return path;
  }

  // Create an array of point tuples, [prev, current, next]
  const pairs = rotate(windowsOf(curvePoints, 3, true), -1);

  const curveInfoList: CurveInfo[] = pairs.map((pair) => {
    const [prev, current, next] = pair;

    const currentPoint = current.point;
    const nextPoint = next.point;
    const prevPoint = prev.point;

    const radius = getCornerRadius(
      prevPoint,
      currentPoint,
      nextPoint,
      current.cornerRadius,
    );

    const { step, drawSegment } = getStep(currentPoint, nextPoint, radius);

    return { radius, step, drawSegment };
  });

  zip(pairs, curveInfoList).forEach(([pair, curveInfo], index) => {
    const [, current, next] = pair;
    const isLast = index === pairs.length - 1;

    const currentPoint = current.point;
    const nextPoint = next.point;

    if (index === 0) {
      const { step } = curveInfo;

      path.moveTo(currentPoint.x + step.x, currentPoint.y + step.y);

      if (next.curveMode !== Sketch.CurveMode.Straight) {
        curveTo(path, current, next);
      }
    } else {
      if (
        // TODO: Handle one straight point with radius and one curve
        current.curveMode === Sketch.CurveMode.Straight &&
        next.curveMode === Sketch.CurveMode.Straight
      ) {
        stepToCurve(path, current, curveInfo);

        const { step, drawSegment } = curveInfo;

        if (drawSegment && (isClosed || !isLast)) {
          path.lineTo(nextPoint.x - step.x, nextPoint.y - step.y);
        }
      } else if (isClosed || !isLast) {
        curveTo(path, current, next);
      }
    }

    if (isClosed && isLast) {
      const [, current, next] = pairs[0];

      // If we have a corner radius, draw the final curve
      if (
        current.curveMode === Sketch.CurveMode.Straight &&
        next.curveMode === Sketch.CurveMode.Straight
      ) {
        stepToCurve(path, current, curveInfoList[0]);
      }

      // We need the path to be closed for line caps to draw correctly
      path.close();
    }
  });

  return path;
}

const getBorderPositionPathOp = (
  CanvasKit: CanvasKit,
  position: Sketch.BorderPosition,
): PathOp | undefined => {
  switch (position) {
    case Sketch.BorderPosition.Center:
      return;
    case Sketch.BorderPosition.Outside:
      return CanvasKit.PathOp.Difference;
    case Sketch.BorderPosition.Inside:
      return CanvasKit.PathOp.Intersect;
  }
};

export function getStrokedPath(
  CanvasKit: CanvasKit,
  path: Path,
  width: number,
  lineCap: Sketch.LineCapStyle,
  lineJoin: Sketch.LineJoinStyle,
  pathOp?: PathOp,
): Path {
  const copy = path.copy();

  // A 0-width path will never change the original when unioned
  if (width === 0 && pathOp === CanvasKit.PathOp.Union) return copy;

  if (
    !copy.stroke({
      width,
      cap: lineCapStyle(CanvasKit, lineCap),
      join: lineJoinStyle(CanvasKit, lineJoin),
    })
  ) {
    console.info('[getStrokedPath] Failed to stroke path');
    return copy;
  }

  if (pathOp !== undefined) {
    if (!copy.op(path, pathOp)) {
      console.info('[getStrokedPath] Failed to combine paths');
    }
  }

  return copy;
}

export function getStrokedBorderPath(
  CanvasKit: CanvasKit,
  path: Path,
  borderWidth: number,
  borderPosition: Sketch.BorderPosition,
  lineCap: Sketch.LineCapStyle,
  lineJoin: Sketch.LineJoinStyle,
) {
  switch (borderPosition) {
    case Sketch.BorderPosition.Center:
      break;
    case Sketch.BorderPosition.Outside:
    case Sketch.BorderPosition.Inside:
      borderWidth *= 2;
  }

  const pathOp = getBorderPositionPathOp(CanvasKit, borderPosition);

  return getStrokedPath(
    CanvasKit,
    path,
    borderWidth,
    lineCap,
    lineJoin,
    pathOp,
  );
}

export function unscaleCurvePoint(curvePoint: Sketch.CurvePoint, frame: Rect) {
  const unscale = (value: string) =>
    PointString.encode(unscalePoint(PointString.decode(value), frame));

  return {
    ...curvePoint,
    curveFrom: unscale(curvePoint.curveFrom),
    curveTo: unscale(curvePoint.curveTo),
    point: unscale(curvePoint.point),
  };
}

function pathCommandToSVGCommand(
  command: PathCommand,
): CommandWithoutQuadratics {
  switch (command[0]) {
    case PathCommandVerb.move: {
      const [, x, y] = command;
      return { type: 'move', to: { x, y } };
    }
    case PathCommandVerb.cubic: {
      const [, x1, y1, x2, y2, x3, y3] = command;
      return {
        type: 'cubicCurve',
        to: { x: x3, y: y3 },
        controlPoint1: { x: x1, y: y1 },
        controlPoint2: { x: x2, y: y2 },
      };
    }
    default:
      throw new Error('Path command not handled');
  }
}

export function pathToCurvePoints(
  path: Path,
  frame: Rect,
): Sketch.CurvePoint[] {
  const svgCommands = parsePathCmds(path.toCmds()).map(pathCommandToSVGCommand);

  // Assume a single path. `isClosed` should already be handled, before calling
  // this function, so we can ignore it here.
  const curvePoints = makePathsFromCommands(svgCommands)[0].points;

  return curvePoints.map((curvePoint) => unscaleCurvePoint(curvePoint, frame));
}

export function joinCurvePoints(
  segments: Sketch.CurvePoint[][],
  wrapsAround: boolean,
): Sketch.CurvePoint[] {
  const nonEmptySegments = segments.filter((segment) => segment.length > 0);

  if (nonEmptySegments.length === 0) return [];

  const [first, ...rest] = nonEmptySegments;

  const merged = rest.reduce((start, end) => {
    const lastStartPoint = start[start.length - 1];
    const firstEndPoint = end[0];

    return [
      ...start.slice(0, -1),
      SketchModel.curvePoint({
        curveMode: Sketch.CurveMode.Mirrored,
        hasCurveFrom: true,
        hasCurveTo: true,
        point: lastStartPoint.point,
        curveFrom: firstEndPoint.curveFrom,
        curveTo: lastStartPoint.curveTo,
      }),
      ...end.slice(1),
    ];
  }, first);

  if (wrapsAround) {
    const lastIndex = merged.length - 1;
    const last = merged[lastIndex];
    merged.splice(lastIndex, 1);
    merged[0].curveTo = last.curveTo;
  }

  return merged;
}

export function splitPath(path: Path, t: number): [Path, Path] {
  t = clamp(t, 0, 1);

  return [path.copy().trim(0, t, false)!, path.copy().trim(t, 1, false)!];
}
