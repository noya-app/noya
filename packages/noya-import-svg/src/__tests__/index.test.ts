import fs from 'fs';
import { debugDescription } from 'noya-sketch-model';
import path from 'path';
import { svgToLayer } from '..';

const circleSvg = fs.readFileSync(path.join(__dirname, 'circle.svg'), 'utf8');
const demoSvg = fs.readFileSync(path.join(__dirname, 'demo.svg'), 'utf8');
const bowtieSvg = fs.readFileSync(path.join(__dirname, 'bowtie.svg'), 'utf8');
const bowtieViewBox = fs.readFileSync(
  path.join(__dirname, 'bowtie-viewbox.svg'),
  'utf8',
);

test('makes layers', () => {
  const rootLayer = svgToLayer(circleSvg);

  expect(
    debugDescription(rootLayer, {
      style: true,
      fills: true,
      borders: true,
      points: true,
    }),
  ).toMatchSnapshot();
});

test('makes demo layers', () => {
  const rootLayer = svgToLayer(demoSvg);

  expect(debugDescription(rootLayer)).toMatchSnapshot();
});

test('makes multiple shapePaths in a shapeGroup', () => {
  const rootLayer = svgToLayer(bowtieSvg);

  expect(debugDescription(rootLayer, { points: true })).toMatchSnapshot();
});

test('uses viewbox', () => {
  const rootLayer = svgToLayer(bowtieViewBox);

  expect(debugDescription(rootLayer, { points: true })).toMatchSnapshot();
});
