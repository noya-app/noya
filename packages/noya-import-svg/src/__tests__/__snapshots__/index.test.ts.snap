// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makes demo layers 1`] = `
"<Group> svg { x: 0, y: 0, w: 517, h: 517.96 }
├── <ShapeGroup> Path { x: 0, y: 290.75, w: 133.82, h: 92.42 }
│   └── <ShapePath> Path { x: 0, y: 0, w: 133.82, h: 92.42 }
├── <ShapeGroup> Path { x: 0, y: 290.75, w: 133.82, h: 92.42 }
│   ├── <ShapePath> Path { x: 0, y: 0, w: 133.82, h: 92.42 }
│   └── <ShapePath> Path { x: 0, y: 0, w: 133.82, h: 92.42 }
├── <ShapeGroup> Path { x: 172.33, y: 290.75, w: 133.82, h: 92.42 }
│   └── <ShapePath> Path { x: 0, y: 0, w: 133.82, h: 92.42 }
├── <ShapeGroup> Path { x: 162.71, y: 281.12, w: 153.08, h: 111.68 }
│   ├── <ShapePath> Path { x: 0, y: 0, w: 153.08, h: 111.68 }
│   └── <ShapePath> Path { x: 0, y: 0, w: 153.08, h: 111.68 }
├── <ShapeGroup> Path { x: 363.92, y: 290.75, w: 133.82, h: 92.42 }
│   └── <ShapePath> Path { x: 0, y: 0, w: 133.82, h: 92.42 }
├── <ShapeGroup> Path { x: 344.67, y: 271.5, w: 172.33, h: 130.93 }
│   ├── <ShapePath> Path { x: 0, y: 0, w: 172.33, h: 130.93 }
│   └── <ShapePath> Path { x: 0, y: 0, w: 172.33, h: 130.93 }
├── <ShapeGroup> Path { x: 344.67, y: 0, w: 143.45, h: 139.6 }
│   └── <ShapePath> Path { x: 0, y: 0, w: 143.45, h: 139.6 }
├── <ShapeGroup> Path { x: 344.67, y: 0, w: 143.45, h: 139.6 }
│   ├── <ShapePath> Path { x: 0, y: 0, w: 143.45, h: 139.6 }
│   └── <ShapePath> Path { x: 0, y: 0, w: 143.45, h: 139.6 }
├── <ShapeGroup> Path { x: 158.85, y: 421.69, w: 96.28, h: 96.28 }
│   └── <ShapePath> Path { x: 0, y: 0, w: 96.28, h: 96.28 }
├── <ShapeGroup> Path { x: 274.39, y: 421.69, w: 96.28, h: 96.28 }
│   └── <ShapePath> Path { x: 0, y: 0, w: 96.28, h: 96.28 }
└── <ShapeGroup> Path { x: 389.92, y: 421.69, w: 96.28, h: 96.28 }
    └── <ShapePath> Path { x: 0, y: 0, w: 96.28, h: 96.28 }"
`;

exports[`makes layers 1`] = `
"<Group> svg { x: 0, y: 0, w: 90, h: 90 }
├── <Style>
└── <ShapeGroup> Path { x: 0, y: 0, w: 90, h: 90 }
    ├── <Style>
    │   ├── <Fill> (Color)
    │   │   └── <Color> { red: 0, green: 0.47, blue: 1, alpha: 1 }
    │   └── <Border> (Color)
    │       └── <Color> { red: 0, green: 0, blue: 0, alpha: 1 }
    └── <ShapePath> Path { x: 0, y: 0, w: 90, h: 90 }
        ├── <Style>
        ├── <CurvePoint> { x: 1, y: 0.5 } (Mirrored)
        ├── <CurvePoint> { x: 0.5, y: 1 } (Mirrored)
        ├── <CurvePoint> { x: 0, y: 0.5 } (Mirrored)
        ├── <CurvePoint> { x: 0.5, y: 0 } (Mirrored)
        └── <CurvePoint> { x: 1, y: 0.5 } (Mirrored)"
`;

exports[`makes multiple shapePaths in a shapeGroup 1`] = `
"<Group> svg { x: 0, y: 0, w: 200, h: 200 }
└── <ShapeGroup> Path { x: 0, y: 0, w: 200, h: 200 }
    ├── <ShapePath> Path { x: 0, y: 0, w: 200, h: 200 }
    │   ├── <CurvePoint> { x: 0.5, y: 1 } (Straight)
    │   ├── <CurvePoint> { x: 0, y: 0.5 } (Straight)
    │   └── <CurvePoint> { x: 0.5, y: 0.5 } (Straight)
    └── <ShapePath> Path { x: 0, y: 0, w: 200, h: 200 }
        ├── <CurvePoint> { x: 0.5, y: 0 } (Straight)
        ├── <CurvePoint> { x: 1, y: 0.5 } (Straight)
        └── <CurvePoint> { x: 0.5, y: 0.5 } (Straight)"
`;

exports[`uses viewbox 1`] = `
"<Group> svg { x: 0, y: 0, w: 80, h: 80 }
└── <ShapeGroup> Path { x: 0, y: 0, w: 80, h: 80 }
    ├── <ShapePath> Path { x: 0, y: 0, w: 80, h: 80 }
    │   ├── <CurvePoint> { x: 0.5, y: 1 } (Straight)
    │   ├── <CurvePoint> { x: 0, y: 0.5 } (Straight)
    │   └── <CurvePoint> { x: 0.5, y: 0.5 } (Straight)
    └── <ShapePath> Path { x: 0, y: 0, w: 80, h: 80 }
        ├── <CurvePoint> { x: 0.5, y: 0 } (Straight)
        ├── <CurvePoint> { x: 1, y: 0.5 } (Straight)
        └── <CurvePoint> { x: 0.5, y: 0.5 } (Straight)"
`;
